code/parachain/frame/cosmwasm/src/ibc.rs

```diff
@@ -0,0 +1,90 @@
+use alloc::string::{String, ToString};
+use sp_std::{marker::PhantomData, str::FromStr};
+
+use frame_support::{ensure, RuntimeDebug};
+use ibc::core::{
+	ics04_channel::{
+		channel::{Counterparty, Order},
+		error::Error as IbcError,
+		Version,
+	},
+	ics24_host::identifier::{ChannelId, ConnectionId, PortId},
+	ics26_routing::context::{Module as IbcModule, ModuleId, ModuleOutputBuilder},
+};
+use pallet_ibc::routing::ModuleRouter as IbcModuleRouter;
+
+use crate::{Config, Pallet};
+
+#[derive(Default, RuntimeDebug, Eq, PartialEq, Clone)]
+pub struct Router<T: Config> {
+	_marker: PhantomData<T>,
+}
+
+impl<T: Config + Send + Sync> IbcModule for Router<T> {
+	fn on_chan_open_init(
+		&mut self,
+		_output: &mut ModuleOutputBuilder,
+		order: Order,
+		_connection_hops: &[ConnectionId],
+		port_id: &PortId,
+		_channel_id: &ChannelId,
+		_counterparty: &Counterparty,
+		_version: &Version,
+	) -> Result<(), IbcError> {
+		let port_contract =
+			<Pallet<T>>::cosmwasm_addr_to_account(port_id.to_string()).map_err(|_| {
+				IbcError::implementation_specific("contract for port not found".to_string())
+			})?;
+
+		// this check to be moved into contract
+		ensure!(
+			order == Order::Unordered,
+			IbcError::unknown_order_type("can work only with Unordered channels".to_string())
+		);
+		if let Ok(_contract) = <Pallet<T>>::contract_info(&port_contract) {
+			log::trace!(target: "cosmwasm::ibc", "contract for port found");
+			// here we should call `ibc_channel_open`
+			Ok(())
+		} else {
+			Err(IbcError::implementation_specific(
+				"contract for desired port not found".to_string(),
+			))
+		}
+	}
+
+	fn on_chan_open_try(
+		&mut self,
+		_output: &mut ModuleOutputBuilder,
+		_order: Order,
+		_connection_hops: &[ConnectionId],
+		_port_id: &PortId,
+		_channel_id: &ChannelId,
+		_counterparty: &Counterparty,
+		_version: &Version,
+		_counterparty_version: &Version,
+	) -> Result<Version, IbcError> {
+		// here we should call `ibc_channel_open`
+		Err(IbcError::implementation_specific("unimplemented!".to_string()))
+	}
+}
+
+impl<T: Config + Send + Sync + Default> IbcModuleRouter for Router<T> {
```

Comment:
It would also be better to implement this in the runtime module like it's done here https://github.com/ComposableFi/centauri/blob/master/utils/parachain-node/runtime/src/lib.rs#L535-L558

Reply:
That is the same. We are not going to have any router then this. If the future can have runtime meta router routing to routers. So likely cosmwasm router will barely be changed to be one instead of only. 

Good improvement, but can be done later.

Reply:
BTW, `reflect` contract is better ping :)

Reply:
I am aware of the slowness of the contracts, but that would be premature optimization IMHO for doing too much in the substrate.

Reply:
Again, that is XCVM team work, I am not aware of bridge product plans.

Reply:
will not do ping. for ics20 router in runtime see other comment. there will be 2 impl for routers and one composite tuple.

Reply:
(similar to what is in XCM)